File: runtimes/go/pkg/function_async.go
package pkg

import (
	"encoding/json"
	"math/rand"
)

func randomWord() string {
	words := []string{"apple", "banana", "cherry", "date", "elderberry"}
	return words[rand.Intn(len(words))]
}

type Payload struct {
	Word string `json:"word"`
}

func StreamHandler(input <-chan []byte) <-chan []byte {
	out := make(chan []byte)

	go func() {
		defer close(out)

		for range input {
			data := Payload{
				Word: randomWord(),
			}

			jsonBytes, err := json.Marshal(data)
			if err != nil {
				continue // safer than panic in a stream
			}

			out <- jsonBytes
		}
	}()

	return out
}

File: runtimes/rust/pkg/function_async.rs
use rand::seq::SliceRandom;
use serde::Serialize;
use tokio::sync::mpsc::{self, Receiver};

#[derive(Serialize)]
struct Payload {
    word: String,
}

fn random_word() -> String {
    let words = ["apple", "banana", "cherry", "date", "elderberry"];
    words
        .choose(&mut rand::thread_rng())
        .unwrap_or(&"apple")
        .to_string()
}

pub fn stream_handler(mut input: Receiver<Vec<u8>>) -> Receiver<Vec<u8>> {
    let (tx, rx) = mpsc::channel(16);

    tokio::spawn(async move {
        while input.recv().await.is_some() {
            let payload = Payload {
                word: random_word(),
            };
            if let Ok(json_bytes) = serde_json::to_vec(&payload) {
                if tx.send(json_bytes).await.is_err() {
                    break;
                }
            }
        }
    });

    rx
}
